crates/cadenza-eval/src/special_form.rs

```diff
@@ -0,0 +1,142 @@
+///
+/// This replaces the previous `BuiltinMacro` concept, which only handled evaluation.
+pub trait SpecialForm: Send + Sync {
+    /// Returns the name of this special form for display/debugging.
+    fn name(&self) -> &'static str;
+
+    /// Returns the type signature of this special form.
+    fn signature(&self) -> Type;
+
+    /// Evaluates this special form with unevaluated AST arguments.
+    ///
+    /// The special form receives the raw AST expressions and can choose which
+    /// to evaluate and when. This is what makes special forms "special" - they
+    /// control evaluation rather than receiving pre-evaluated arguments.
+    ///
+    /// # Arguments
+    /// - `args`: Unevaluated AST expressions passed to the special form
+    /// - `ctx`: Evaluation context providing access to environment and compiler
+    ///
+    /// # Returns
+    /// The evaluated result as a runtime Value, or a diagnostic error.
+    fn eval(&self, args: &[Expr], ctx: &mut EvalContext<'_>) -> Result<Value>;
+
+    /// Generates IR instructions for this special form.
+    ///
+    /// The special form receives the raw AST expressions and generates appropriate
+    /// IR instructions. It can generate IR for sub-expressions by calling the
+    /// provided gen_expr function.
+    ///
+    /// # Arguments
+    /// - `args`: Unevaluated AST expressions passed to the special form
+    /// - `block`: IR block builder for emitting instructions
+    /// - `ctx`: IR generation context for variable bindings and types
+    /// - `source`: Source location information for debugging
+    /// - `gen_expr`: Function to generate IR for a sub-expression
+    ///
+    /// # Returns
+    /// The ValueId of the result in the IR, or an error message.
+    fn build_ir(
+        &self,
+        args: &[Expr],
+        block: &mut BlockBuilder,
+        ctx: &mut IrGenContext,
+        source: SourceLocation,
+        gen_expr: &mut dyn FnMut(&Expr, &mut BlockBuilder, &mut IrGenContext) -> std::result::Result<ValueId, String>,
+    ) -> std::result::Result<ValueId, String>;
+}
+
+/// A special form implementation with function pointers.
+///
+/// This is the concrete type used for most built-in special forms.
+/// It stores function pointers for both evaluation and IR generation.
+pub struct BuiltinSpecialForm {
+    /// The special form name for display/debugging.
+    pub name: &'static str,
+    /// The type signature of this special form.
+    pub signature: Type,
+    /// The evaluation function (receives unevaluated AST expressions).
+    pub eval_fn: fn(&[Expr], &mut EvalContext<'_>) -> Result<Value>,
+    /// The IR generation function (generates IR from AST expressions).
+    pub ir_fn: fn(
+        &[Expr],
+        &mut BlockBuilder,
+        &mut IrGenContext,
+        SourceLocation,
+        &mut dyn FnMut(&Expr, &mut BlockBuilder, &mut IrGenContext) -> std::result::Result<ValueId, String>,
+    ) -> std::result::Result<ValueId, String>,
```

Comment:
Is there a reason we're not using diagnostics here for errors? We really should be.